sdk/go/docker/customTypes.go

```diff
@@ -0,0 +1,266 @@
+// Copyright 2016-2020, Pulumi Corporation.
+
+package docker
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/pulumi/pulumi/sdk/go/pulumi"
+)
+
+type CacheFrom struct {
```

Comment:
This does get more complicated with nested types. I'd love to avoid adding yet another code generator, but if we did, we'd probably do something like

```
//+pulumi
type Foo struct {
}
```

And then let a code generator chew on that. We might be able to do this by refactoring the schema generator a bit and adding a frontend that works off of parsed code.

At the moment, though, only the structs with input-typed fields should be necessary, and they shouldn't need to implement `Input`. The result will be a bit clumsy to work with--it will need type assertions--but it does avoid a fair bit of hand-written code.